prow/cmd/branchprotector/protect.go

```diff
@@ -0,0 +1,286 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/url"
+	"strings"
+	"sync"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+)
+
+type options struct {
+	config   string
+	token    string
+	confirm  bool
+	endpoint string
+}
+
+func jobRequirements(jobs []config.Presubmit, after bool) []string {
+	var required []string
+	for _, j := range jobs {
+		if !after && (!j.AlwaysRun || j.SkipReport) {
+			continue
+		}
+		required = append(required, j.Context)
+		required = append(required, jobRequirements(j.RunAfterSuccess, true)...)
+	}
+	return required
+}
+
+func repoRequirements(org, repo string, cfg config.Config) []string {
+	p, ok := cfg.Presubmits[org+"/"+repo]
+	if !ok {
+		return nil
+	}
+	return jobRequirements(p, false)
+}
+
+func flagOptions() options {
+	o := options{}
+	flag.StringVar(&o.endpoint, "endpoint", "https://api.github.com", "Github api endpoint, may differ for enterprise")
+	flag.StringVar(&o.config, "config", "", "Path to prow config.yaml")
+	flag.StringVar(&o.token, "token", "", "Path to github token")
```

Comment:
[`github-token-file`](https://github.com/kubernetes/test-infra/blob/83c90ff53262812ab288afc7188201cc8ffa35fd/prow/cmd/plank/main.go#L51) to match what is used elsewhere. There is also a default we set to `/etc/github/oauth`.

Reply:
Changed to github-token-file (let it be known for posterity that I consider it unnecessarily grotesquely redundant and verbose)

Leaving the default value empty (which I think we should change elsewhere as it makes the yaml files easier to read and understand... why am I mapping something to /etc/github/oauth?)

Reply:
Used as a default everywhere else we use this flag and by now it seems like a convention for our tools.

> which I think we should change elsewhere as it makes the yaml files easier to read and understand...

I think that the default is there in the first place so that the resulting YAML will be less verbose thus easier to read. I don't feel strong either way but I don't expect most people to have opinions about where to place stuff like the github or hmac token, or the prow config file (99% of the times these binaries are deployed are as part of an automated process in a container anyway).